src/main.rs

```diff
@@ -61,34 +59,52 @@ enum EmitArgs {
 
 lalrpop_mod!(pub rascal_grammar);
 
-fn main() {
-    let args = Args::parse();
-    let save_c: bool;
-    let save_ir: bool;
-    if let Some(emit) = args.emit {
-        save_c = emit.iter().any(|x| matches!(x, EmitArgs::C));
-        save_ir = emit.iter().any(|x| matches!(x, EmitArgs::Ir));
-    } else {
-        (save_c, save_ir) = (false, false);
-    }
-    let src_file = fs::read_to_string(&args.infile).expect("ERROR: couldn't find source file");
-    let build_stack = if Path::new(&args.infile).extension().expect("Problem with filename") == "ir" {
+#[derive(Error, Debug)]
+pub enum BuildError {
+    #[error("There was an issue with the input file: {0}")]
+    Input(String),
+    #[error("There was a problem creating the output: {0}")]
+    Output(String),
+}
+
+fn build_stack_from_input(infile: &String, save_ir: bool) -> Result<Vec<IRNode>, BuildError> {
```

Comment:
Overall I like the new `BuildError` type for handling these types of errors. Could you try to handle the error without using this separate function though? Creating the build stack will probably end up being the majority of the code for the compiler so I just want to separate errors that come from the checking and type inference part and errors that come from extraneous cases like "file not found" issues? For instance maybe `main` can return a `Result<()>`? Sort of like [this](https://docs.rs/anyhow/latest/anyhow/#details)? I'm not describing this super well sorry. But I also want to avoid calling `.expect()` in main. While it creates a succinct way to error, I think it would be nice to explicity return an exit code, or get this for free by return a result. What do you think? Hopefully that makes sense lol

Reply:
check this shit out